block-scoped-var gives a false positive on function argument destructuring

```shell
$ eslint --version
v0.16.1
```

.eslintrc

```json
{
  "env": {
    "es6": true
  },
  "rules": {
    "block-scoped-var": 2
  },
  "ecmaFeatures": {
    "destructuring": true
  }
}
```

test.js
```js
'use strict';

var fun = function({x}) {
  return x;
};

var fun2 = function(arg) {
  var {x} = arg;
  return x;
};

fun({x: 2});
fun2({x: 2});
```

Error
```
test.js
  4:9  error  "x" used outside of binding context  block-scoped-var

✖ 1 problem (1 error, 0 warnings)
```

As you can see, the second function is linted correctly (due to #1863), but the first one fails.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9399824-block-scoped-var-gives-a-false-positive-on-function-argument-destructuring?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/lib/rules/block-scoped-var.js
+++ b/lib/rules/block-scoped-var.js
@@ -89,9 +89,31 @@ module.exports = function(context) {
 
     function functionHandler(node) {
         pushScope();
-        declare(node.params.map(function(id) {
-            return id.name;
-        }));
+
+        node.params.forEach(function(param) {
+
+            switch (param.type) {
+                case "Identifier":
+                    declare([param.name]);
+                    break;
+
+                case "ObjectPattern":
+                    declare(param.properties.map(function(property) {
+                        return property.key.name;
+                    }));
+                    break;
+
+                case "ArrayPattern":
+                    declare(param.elements.map(function(element) {
+                        return element.name;
+                    }));
+                    break;
+
+                // no default
+            }
+
+        });
+
         declare(node.id ? [node.id.name] : []);
         declare(node.rest ? [node.rest.name] : []);
         declare(["arguments"]);

--- a/tests/lib/rules/block-scoped-var.js
+++ b/tests/lib/rules/block-scoped-var.js
@@ -38,6 +38,7 @@ eslintTester.addRuleTest("lib/rules/block-scoped-var", {
         { code: "new Date", globals: {Date: false} },
         { code: "new Date", globals: {} },
         { code: "var eslint = require('eslint');", globals: {require: false} },
+        { code: "var fun = function({x}) {return x;};", ecmaFeatures: { destructuring: true } },
         "function f(a) { return a.b; }",
         "var a = { \"foo\": 3 };",
         "var a = { foo: 3 };",

